feat: Added `SettingPermissions` struct and HCL schema

d0weinberger opened the pull request:
This PR introduces the structure of the `dynatrace_setting_permissions` resource and its HCL representation. Accordingly, `Schema()`, `MarshalHCL()` and `UnmarshalHCL()` functions are implemented, and an example config is given in `example_a.tf`.

In the following example, a permission resource is defined for the settings object ID of another Terraform resource `example_connection` of type `dynatrace_github_connection`.

It specifically grants **no access** to arbitrary users.
The user `user_id_1` gets write access and the user `user_id_2` gets read access.
Additionally, users of group `group_id_2` get write access.

Permissions for groups and users not defined in the resource will be **revoked**!

```
resource "dynatrace_setting_permissions" "github_connection_access" {
  settings_object_id = dynatrace_github_connection.example_connection.id
  all_users = "none" # possible values: none, read, write - default: none
  users { # default: empty
    user {
        user_id = "user_id_1"
        access = "write" # possible values: read, write
    }
    user {
        user_id = "user_id_2"
        access = "read" # possible values: read, write
    }
  }
  groups { # default: empty
    group {
        group_id = "group_id_2"
        access = "write" # possible values: read, write
    }
  }
}
```

d0weinberger committed:
feat: Added SettingPermissions schema and example

Kirdock commented on dynatrace/api/v2/settings/objects/permissions/settings/owner_based_access_control.go:
```diff
@@ -0,0 +1,71 @@
+	SettingsObjectID string `json:"settings_object_id"`
+	AllUsers         string `json:"all_users"`
+	Users            Users  `json:"users"`
+	Groups           Groups `json:"groups"`
```
Do we even need JSON tags for this one? We won't marshal this one as these are separate endpoints right?

d0weinberger replied:
I guess you're right. We could get rid of the tags. The structs which will be marshalled aren't defined yet.

d0weinberger replied:
Done

Kirdock commented on dynatrace/api/v2/settings/objects/permissions/settings/user.go:
```diff
@@ -0,0 +1,79 @@
+
+type UserAccessor struct {
+	UserID string `json:"user_id"`
+	Access string `json:"access"`
```
Is there something else I'm not seeing at the moment (mapper function or another struct)?
I can't see the direct mapping from HCL to the API, which should be part of CA-15663, right?

This resource should be marshalled to the API request, which is
```json
{
  "accessor": {
    "id": "abc-xy",
    "type": "user"
  },
  "permissions": [
    "r"
  ]
}
```

d0weinberger replied:
Oh, that is not yet done, no. This is really just the struct representation of the HCL.

Kirdock commented on dynatrace/api/v2/settings/objects/permissions/settings/group.go:
```diff
@@ -0,0 +1,79 @@
+		"access": {
+			Type:        schema.TypeString,
+			Description: "Valid values: read, write",
+			Required:    true,
```
Should we add a validation function here and also for other resources?
```suggestion
			Required:    true,
			ValidateFunc: validation.StringInSlice([]string{"read", "write"}, false),
```

d0weinberger replied:
Validation makes sense, yes. I will add that

d0weinberger replied:
Done

tomazpu commented on dynatrace/api/v2/settings/objects/permissions/testdata/terraform/example_a.tf:
```diff
@@ -0,0 +1,20 @@
+        access = "write"
+    }
+  }
+}
```
New line missing at end of file

d0weinberger replied:
Done

tomazpu commented on dynatrace/api/v2/settings/objects/permissions/testdata/terraform/example_a.tf:
```diff
@@ -0,0 +1,20 @@
+resource "dynatrace_owner_based_access_control" "ABC" {
+  settings_object_id = "asdf"
```
If this value is in the documentation, I would prefer something like `"#id#"`

Laubi replied:
Or we could reference some other fake object to show how we'd recommend using it?

d0weinberger replied:
I would like to use `dynatrace_github_connection.example_connection.id` but I am not sure "id" is the id of the settings object here.

EDIT: Okay, I'm a bit more confident now that this works: https://github.com/dynatrace-oss/terraform-provider-dynatrace/blob/3202ca45f15bcb7869a19ae2b6728de0893d7285/dynatrace/settings/services/settings20/service.go#L463

d0weinberger replied:
Done

tomazpu commented on dynatrace/api/v2/settings/objects/permissions/testdata/terraform/example_a.tf:
```diff
@@ -0,0 +1,20 @@
+resource "dynatrace_owner_based_access_control" "ABC" {
```
Instead of "ABC" i would prefer "#name#"

d0weinberger replied:
I will name it `github_connection_access`

d0weinberger replied:
Done

tomazpu commented on dynatrace/api/v2/settings/objects/permissions/settings/group.go:
```diff
@@ -0,0 +1,79 @@
+/**
```
I would probably prefer the file be named `group_accessor` to prevent confusion about the group's undescriptiveness.

d0weinberger replied:
I will rename `user` to `user_accessor` as well

d0weinberger replied:
Done

Kirdock commented on dynatrace/api/v2/settings/objects/permissions/testdata/terraform/example_a.tf:
```diff
@@ -0,0 +1,20 @@
+resource "dynatrace_setting_permissions" "github_connection_access" {
+  settings_object_id = dynatrace_github_connection.example_connection.id
+  all_users = "write"
```
[nit] should we change it to `read` for this one? Not sure if this is still the case but `allUsers: write` would override/delete the others, right?
Or we could change it in a way so that all the other arguments here make sense? Not sure which accessor type has which priority, but I guess the most powerful one is taken?

Kirdock commented on dynatrace/api/v2/settings/objects/permissions/settings/setting_permissions.go:
```diff
@@ -0,0 +1,73 @@
+			Required: true,
+		},
+		"all_users": {
+			Type:         schema.TypeString,
```
At the moment, the description is a bit inconsistent. Some have one, some haven't, and for some it's `No documentation available`. Should we improve this here?

d0weinberger replied:
I'll mention that it's the UUID in Dynatrace and maybe add an example to use the data source

Kirdock commented on dynatrace/api/v2/settings/objects/permissions/settings/setting_permissions.go:
```diff
@@ -0,0 +1,73 @@
+}
+
+func (_ *SettingPermissions) Schema() map[string]*schema.Schema {
+	return map[string]*schema.Schema{
```
We should also add descriptions for these arguments.
I'm not sure about the group arguments, though (`groups` and `users`)

d0weinberger replied:
Done

Kirdock commented on dynatrace/api/v2/settings/objects/permissions/testdata/terraform/example_a.tf:
```diff
@@ -0,0 +1,32 @@
+data "dynatrace_iam_user" "user_a" {
+  email = "a@home.com"
```
Maybe we should change this one to `example.com` or even `dynatrace.com`. We should not forward to any other sites (`home.com` redirects to `fairway.com`)

d0weinberger replied:
Changed this and several other occurrences to `example.com`

d0weinberger committed:
feat: Replacing occurrences of "@home.com" with "@example.com"

d0weinberger committed:
feat: Adding descriptions to group and user schemas.

d0weinberger committed:
feat: Changing types for "user" and "group"

Types for the schemas "user" and "group" are changed from "schema.TypeList" to "schema.TypeSet", as the order is not important.

d0weinberger committed:
feat: Adding MaxItems to "users" and "groups" schemas

To allow "user" blocks nested inside "users", the "users" schema needs to have a type like, e.g. "schema.TypeList". To prevent the creation of several "users" blocks, we need to set "MaxItems: 1". The same is true for the "groups" schema.

d0weinberger committed:
feat: descriptions for settings_object_id and all_users

Kirdock commented on dynatrace/api/v2/settings/objects/permissions/settings/user_accessor.go:
```diff
@@ -0,0 +1,81 @@
+
+func (_ *UserAccessor) Schema() map[string]*schema.Schema {
+	return map[string]*schema.Schema{
+		"user_id": {
```
Sry, for the late feedback: Thinking about it again, should we just call it `user` or `user_uid`? The API and our resource return it as `uid`.

When looking at other resources like `policy_binding`, we don't attach any identifier to the arguments and just call them `environment/account/policies`, but I also see that this is a bit inconsistent (different for management zones, dashboard, and application_id).

On the other hand, when comparing different providers, like AWS, they tend to use `_id`.

We could also leave it like that, but I would first like to know the others' opinions.

d0weinberger replied:
We could also drop the `user` prefix as it it in a `user` block anyway.
Maybe just `id`? With `uid` it would be a bit weird compared with groups, because from the group data source, you would use the `id`. There is no `uid`. But, if we wanted to align it with the datasources, it could potentially be:

```
user {
  uid = data.dynatrace_iam_user.user_a.uid
  access = "write"
}
```
and 

```
group {
  id = data.dynatrace_iam_group.group_b.id
  access = "write"
}
```

Kirdock replied:
We decided to go with `user_uid`, because we also have the user resource, which returns the `uid` attribute.

d0weinberger replied:
Okay. Let's wait for other opinions

Kirdock replied:
Good point. I would go for just `uid`/`id` as the rest is just duplicated information from the parent block

Kirdock commented on dynatrace/api/v2/settings/objects/permissions/settings/setting_permissions.go:
```diff
@@ -0,0 +1,77 @@
+
+func (_ *SettingPermissions) Schema() map[string]*schema.Schema {
+	return map[string]*schema.Schema{
+		"settings_object_id": {
```
Also thinking here again, we should probably discuss this a bit more: As a Terraform user, do I know which settings objects can actually be used here, or what settings objects are in general?
For example, we have the `dynatrace_site_reliability_guardian`, which is in the `Automation` section in the docs. Then we have the `dynatrace_github_connection`, which is part of `Connectors`.
Both of them are settings objects, but is it enough if the documentation of them just links to settings and refers to settings permissions?

Also, other settings schemas may be currently in use, and not all of them support defining permissions.

d0weinberger replied:
I get your point. We could add to the description of "settings_object_id" something like `"To understand which resources are supported, check <insert docs here>"` 

Regarding the mentioning of settings objects ... Yeah, that one is interesting. If we only call it a resource without the context that under the hood it's a settings object, then we are also not talking about settings permissions here, but "resouce permissions" maybe?

Kirdock replied:
Follow up here: We should have good documentation that tells which resource (settings) can be used here.

d0weinberger committed:
feat: Renaming of user_id and group_id

"user_id" is now just "uid" and "group_id" is now just "id"

Kirdock commented on dynatrace/api/v2/settings/objects/permissions/settings/user_accessor.go:
```diff
@@ -0,0 +1,81 @@
+
+func (_ *UserAccessor) Schema() map[string]*schema.Schema {
+	return map[string]*schema.Schema{
+		"id": {
```
Not sure if this will lead to confusion, if the user resource has `uid` and this one is `id`, whereas in the description we call it again `UUID`. The API endpoint returns for both (user and group) `uid`

d0weinberger replied:
> The API endpoint returns for both (user and group) uid

Okay, I was focusing on the TF resources and `dynatrace_iam_group` has no uid attribute, therefore i chose "id" there

arthurpitman commented on dynatrace/api/v2/settings/objects/permissions/testdata/terraform/example_a.tf:
```diff
@@ -0,0 +1,32 @@
+}
+
+resource "dynatrace_setting_permissions" "github_connection_access" {
+  settings_object_id = dynatrace_github_connection.example_connection.id
```
Should the example be fully self-contained, i.e., also provide the `dynatrace_github_connection` we reference here?

d0weinberger replied:
Hm, I could of course add it.

EDIT: added it

arthurpitman commented on dynatrace/api/v2/settings/objects/permissions/settings/setting_permissions.go:
```diff
@@ -0,0 +1,77 @@
+
+func (_ *SettingPermissions) Schema() map[string]*schema.Schema {
+	return map[string]*schema.Schema{
+		"settings_object_id": {
```
nit: We write `settings_object_id` but call our resource `dynatrace_setting_permissions`. I wonder if `dynatrace_settings_permissions` makes more sense (also considering the underlying API)? If need be we can still change this, so this is not something blocking this PR.

d0weinberger replied:
Mhm, good question. I think we will revisit this at the story where everything is wired together, because at that point, we need to give the resource a name. We would then also need to adapt the example

d0weinberger committed:
feat: Adding dynatrace_github_connection to example

arthurpitman commented on dynatrace/api/v2/settings/objects/permissions/settings/user_accessor.go:
```diff
@@ -0,0 +1,81 @@
+
+type Users []*UserAccessor
+
+func (g *Users) Schema() map[string]*schema.Schema {
```
For consistency:
```suggestion
func (_ *Users) Schema() map[string]*schema.Schema {
```

arthurpitman commented on dynatrace/api/v2/settings/objects/permissions/settings/user_accessor.go:
```diff
@@ -0,0 +1,81 @@
+	}
+}
+
+func (g *Users) MarshalHCL(properties hcl.Properties) error {
```
nit:
```suggestion
func (u *Users) MarshalHCL(properties hcl.Properties) error {
```

arthurpitman commented on dynatrace/api/v2/settings/objects/permissions/settings/group_accessor.go:
```diff
@@ -0,0 +1,81 @@
+
+type Groups []*GroupAccessor
+
+func (g *Groups) Schema() map[string]*schema.Schema {
```
nit:
```suggestion
func (_ *Groups) Schema() map[string]*schema.Schema {
```
